File: 10-Queue/programs/ruby/program2.rb
# Implementation of circular queue using array

class CircularQueue
  def initialize(size)
    @queue = Array.new(size)
    @front = -1
    @rear = -1
    @size = size
  end

  def enqueue(item)
    if (@front == 0 && @rear == @size - 1) || (@rear == (@front - 1) % (@size - 1))
      puts "Queue is full"
    elsif @front == -1
      @front = 0
      @rear = 0
      @queue[@rear] = item
    elsif @rear == @size - 1 && @front != 0
      @rear = 0
      @queue[@rear] = item
    else
      @rear = @rear + 1
      @queue[@rear] = item
    end
  end

  def dequeue
    if @front == -1
      puts "Queue is empty"
    end

    item = @queue[@front]
    @queue[@front] = nil

    if @front == @rear
      @front = -1
      @rear = -1
    elsif @front == @size - 1
      @front = 0
    else
      @front = @front + 1
    end

    item
  end

  def front
    @queue[@front]
  end

  def rear
    @queue[@rear]
  end

  def size
    @queue.size
  end

  def is_empty
    @front == -1
  end

  def display
    puts @queue
  end
end

if __FILE__ == $0
  queue = CircularQueue.new(5)
  queue.enqueue(10)
  queue.enqueue(20)
  queue.enqueue(30)
  queue.enqueue(40)
  queue.display
  puts queue.front
  puts queue.rear
  queue.dequeue
  queue.display
  puts queue.size
  puts queue.is_empty
end

File: 10-Queue/programs/cpp/program2.cpp
// Queue implementation using array in cpp

#include<bits/stdc++.h>
using namespace std;

class Queue{
  private:
    int size;
    int front;
    int rear;
    int *Q;
  public:
    Queue(){
      size = 10;
      front = rear = -1;
      Q = new int[this->size];
    }
    Queue(int size){
      this->size = size;
      front = rear = -1;
      Q = new int[this->size];
    }
    void enqueue(int x);
    int dequeue();
    void Display();
};

void Queue::enqueue(int x){
  if(rear == size-1){
    cout << "Queue is full" << endl;
  }else{
    rear++;
    Q[rear] = x;
  }
}

int Queue::dequeue(){
  int x = -1;
  if(front == rear){
    cout << "Queue is empty" << endl;
  }else{
    front++;
    x = Q[front];
  }
  return x;
}

void Queue::Display(){
  for(int i = front+1; i <= rear; i++){
    cout << Q[i] << " ";
  }
  cout << endl;
}

int main(){
  Queue q(5);
  q.enqueue(100);
  q.enqueue(200);
  q.enqueue(300);
  q.Display();
  cout << q.dequeue() << endl;
  q.Display();
  q.enqueue(400);
  q.Display();
  return 0;
}

// Output
// 10 20 30
// 10
// 20 30
// 20 30 40

